File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.cpp
#include <bits/stdc++.h>

using namespace std;

vector<string> solution(vector<string> files) {
    
    struct Node {
        string file;
        string head;
        int number;
        
        Node (string file, string head, string number) 
            : file(file), head(head), number(stoi(number)) {}
    };
    
    vector<Node> nodeArray;
    
    // 1. head, number 구하기
    for (string file : files) {
        
        string head, number;
        int i = 0;
        
        for (; i < file.length(); i++) {
            char ch = file[i];
            
            if ('0' <= ch && ch <= '9') {
                break;
            }
            head += ch;
        }
        
        for (; i < file.length(); i++) {
            char ch = file[i];
            
            if ('0' > ch && ch > '9') {
                break;
            }
            number += ch;
        }
        
        transform(head.begin(), head.end(), head.begin(), ::tolower); // 소문자로 전환
        nodeArray.push_back(Node(file, head, number));
    }
    
    // 2. 주어진 조건대로 정렬하기
    stable_sort(nodeArray.begin(), nodeArray.end(), [](Node a, Node b) {
        if (a.head == b.head) {
            return a.number < b.number;
        }
        return a.head < b.head;
    });
    
    vector<string> answer;
    for (Node node : nodeArray) {
        answer.push_back(node.file);
    }
    return answer;
}

File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.java
import java.util.*;

class Solution {
    
    class Node {
        String file;
        String head;
        int number;
        int index;
        
        public Node (String file, String head, String number, int index) {
            this.file = file;
            this.head = head.toUpperCase();
            this.number = number == null ? 0 : Integer.parseInt(number);
            this.index = index;
        }
    }
    
    public String[] solution(String[] files) {
        
        PriorityQueue<Node> pq = new PriorityQueue<>((a, b) -> {  
            
            if (a.head.equals(b.head)) {
                
                if (a.number == b.number) {
                    return a.index - b.index; // 주어진 순서 유지
                }
                return a.number - b.number; // number - 오름차순
            }
            
            return a.head.compareTo(b.head); // head - 대소문자 구분 없이 사전순
        });
        
        int index = 0;
        for (String file : files) {
            String head = "", number = "", tail = "";
            int i = 0;
            
            // head 구하기
            for (; i < file.length(); i++) {
                
                if ('0' <= file.charAt(i) && file.charAt(i) <= '9') {
                    break;
                }
                head += file.charAt(i);
            }
            
            // number 구하기
            for (; i < file.length(); i++) {
                
                if ('0' > file.charAt(i) || file.charAt(i) > '9') {
                    break;
                }
                number += file.charAt(i);
            }
            
            pq.add(new Node(file, head, number, index));
            index++;
        }
        
        String[] answer = new String[files.length];
        int i = 0;
        while (!pq.isEmpty()) {
            Node curr = pq.poll();
            answer[i++] = curr.file;
        }
        return answer;
    }
}